### PR TITLE
Make GatherParameters more robust

### PR DESCRIPTION
**Pull request by asuhan**
Make the order in which parameters are gathered consistent with the
order of arguments received by the forward method of the module.

### CODE REVIEW

File: torch_xla/csrc/module.cpp [hunk 1]
```diff
@@ -27,20 +27,26 @@ namespace {
...
-  for (auto& param : m.get_parameters()) {
-    values->push_back(param->slot()->toTensor());
-    requires_grad->push_back(true);
-  }
-  for (auto& param : m.get_attributes()) {
-    if (!param->slot()->isTensor()) {
-      continue;
+                      const torch::jit::script::Module& m,
+                      const torch::jit::script::Method* forward) {
+  for (torch::jit::IValue* initial_ivalue : forward->initial_ivalues()) {
```
Comment (dlibenzi) [inline torch_xla/csrc/module.cpp]: This is N^2 with N potentially in the 100s.
I'd walk initial_values() and build an unordered_map, then lookup in parameters and attributes enumerations.

Comment (asuhan) [reply torch_xla/csrc/module.cpp]: Yes, ~10000 iterations once per module initialization. This isn't part of the training loop or anything like that. Even millions of iterations should be fine.

Comment (dlibenzi) [reply torch_xla/csrc/module.cpp]: The code will be 1) Smaller 2) Faster.
Might be a no brainer :)

Comment (asuhan) [reply torch_xla/csrc/module.cpp]: Not sure I see why you'd build a map on `initial_values`, we would need to build two sets on parameters and attributes.

Comment (dlibenzi) [reply torch_xla/csrc/module.cpp]: Something like this (pseudo-code):

```
void GatherParameters(std::vector<at::Tensor>* values,
                      std::vector<bool>* requires_grad,
                      const torch::jit::script::Module& m,
                      const torch::jit::script::Method* forward) {
  std::unordered_set<INITIAL_VALUE_TYPE> initial_values;
  for (torch::jit::IValue* initial_ivalue : forward->initial_ivalues()) {
    initial_values.insert(initial_ivalue);
  }
  for (auto& param : m.get_parameters()) {
    if (initial_values.count(param->slot()) > 0) {
      values->push_back(param->slot()->toTensor());
      requires_grad->push_back(true);
      break;
    }
  }
  for (auto& param : m.get_attributes()) {
    if (initial_values.count(param->slot()) > 0 && param->slot()->isTensor()) {
      values->push_back(param->slot()->toTensor());
      requires_grad->push_back(false);
      break;
    }
  }
  for (const auto& sub : m.get_modules()) {
    GatherParameters(values, requires_grad, *sub->module, forward);
  }
}
```

Comment (asuhan) [reply torch_xla/csrc/module.cpp]: Yeah, this isn't what we want. The order is driven by `forward->initial_ivalues()`, that's what comes from the call. If we visit parameters first, then attributes, we still rely on the implicit assumption that this is the order at the call site, which defeats the purpose of this patch.

Comment (dlibenzi) [reply torch_xla/csrc/module.cpp]: Or maybe even shorter?
```
std::unordered_set<INITIAL_VALUE_TYPE> initial_values(forward->initial_ivalues().begin(), forward->initial_ivalues().end());
```

If forward->initial_ivalues() returns a const&.

Comment (dlibenzi) [reply torch_xla/csrc/module.cpp]: > Yeah, this isn't what we want. The order is driven by `forward->initial_ivalues()`, that's what comes from the call. If we visit parameters first, then attributes, we still rely on the implicit assumption that this is the order at the call site, which defeats the purpose of this patch.

Oh, I see.
There is a better way though, given that constraint.
Think about something which is better than N^2 ;)

Comment (asuhan) [reply torch_xla/csrc/module.cpp]: There's no reason to point out it's N^2, I'm very aware of it. N also happens to be small enough for this to not be an issue at all given we only run this once per the entire training process. That being said, I've addressed it anyway.

Comment (dlibenzi) [reply torch_xla/csrc/module.cpp]: > There's no reason to point out it's N^2, I'm very aware of it. N also happens to be small enough for this to not be an issue at all given that we only run this once per the entire training process. That being said, I've addressed it anyway.

There is no reason to have N^2 semantics for 5 lines of extra code either.
N is not small. Can be hundreds ;)

If you had something like this, you'd have only one lookup:

```
enum ValueType {
  kParameter,
  kAttribute
};
std::unordered_map<torch::jit::IValue*, ValueType> value_types;
...
```

**Pull request merged**